### ISSUE TITLE
S3TransferManger with CRT - Uploading a single large file causes memory issues

### ISSUE THREAD

**Issue by jonathansloman** [aws/aws-sdk-java-v2#4168]
### Describe the bug

When uploading a single very large file (for example, 50GB) the java process uses up all its memory and then crashes out.

We are running a java client within a docker container in EKS,  uploading a file from nfs-attached EFS storage to an s3 bucket.  After about 5% of the transfer, we see memory usage jump to about 1GB, then after 10%, it jumps to 2GB, and so on, until it reaches the limit configured for the pod in Kubernetes, and then it locks up and dies (I'm not sure if the pod itself is dying, or being killed by kubernetes due to failed health checks - nothing is logged).

I've gone as high as setting 8GB memory limit, but this wasn't enough to allow the upload to complete.

I've tried various values for targetThroughput (eg 0.1 or 20) and maxConcurrency (eg null, or 2, or 50), without seeing any difference in the memory usage behaviour.

### Expected Behavior

Expect the transfer to be able to complete using a sensible amount of memory.

### Current Behavior

Java process stops logging, then either dies or is killed due to running out of memory.

In dmesg, we see something like:

Memory cgroup out of memory: Killed process 28619 (java) total-vm:7513612kB, anon-rss:2080236kB, file-rss:22760kB, shmem-rss:0kB, UID:0 pgtables:4764kB oom_score_adj:992

### Reproduction Steps

Sorry, our code is too tightly integrated with other stuff to paste. But in essence - we are create an S3TransferManager like this:

S3CrtAsyncClientBuilder s3AsyncClientBuilder = S3AsyncClient.crtBuilder().region(Region.EU_WEST_1).targetThroughputInGbps(10.0)
                .minimumPartSizeInBytes(5242880).maxConcurrency(10);

S3TransferManager.builder().s3Client(s3AsyncClient.build()).build();

And then doing an uploadObject

It may be key that we are running this in a docker container in an EKS cluster.

### Possible Solution

This is pure speculation, but I wonder if something somewhere (in CRT?) is trying to get how much memory is available, and that this hasn't been done in a docker-friendly way - it may be getting the available memory of the host, rather than the memory available to the docker container it's running in.

### Additional Information/Context

We are using this base docker image to run our java service:

eclipse-temurin:17-jre

### AWS Java SDK version used

2.20.97 (and CRT version 0.22.2)

### JDK version used

openjdk version "17.0.7" 2023-04-18 OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7) OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode, sharing)

### Operating System and version

eclipse-temurin is based on ubuntu

**Comment by jonathansloman** [comment]
one thing to add - we're running java with: -XX:MaxRAMPercentage=100

**Comment by debora-ito** [comment]
I think it's the same behavior reported here: https://github.com/aws/aws-sdk-java-v2/issues/4034.

**Comment by debora-ito** [comment]
@jonathansloman are you using a single or multiple instances of S3TransferManager?

**Comment by jonathansloman** [comment]
Ah yes, I hadn't spotted that issue. It does look like the same thing.

We are creating an S3Client like this: 

`S3Client.builder().build()`

But this one isn't actually used (our code handles multiple paths, some of which need the standard S3Client). We then also create an S3AsyncClient like this:

`S3AsyncClient.crtBuilder().region(Region.EU_WEST_1).targetThroughputInGbps(targetThroughputInGbps)
                .minimumPartSizeInBytes(partSizeBytes).maxConcurrency(maxConcurrency).build()`

And use that to create the S3TransferManager:

`S3TransferManager.builder().s3Client(s3AsyncClient).build()`

Is the suggestions that the initially created S3Client object is causing issues, even though it's not being used?

**Comment by debora-ito** [comment]
> Is the suggestions that the initially created S3Client object is causing issues, even though it's not being used?

We are trying to narrow down the scope of investigation, to see if the memory issue is seen when one single instance of S3Client is created and reused across requests, or multiple S3Clients are created. Using multiple S3Clients, while has some valid use cases, will multiply the resources allocated internally by the SDK.

**Comment by jonathansloman** [comment]
Given we don't use the S3Client after creating it, I'd be surprised if it was having any effect. However if it would be useful to you, I could try refactoring the code to not create it and check that we still have the same memory usage issues.

**Comment by debora-ito** [comment]
@jonathansloman we identified a change in the CRT client that will possibly fix the memory issue with TransferManager upload. 

Will keep this issue updated with the progress on the fix.

**Comment by linzhu-isoftstone** [comment]
@debora-ito I ran into the same issue. It's easily to run into OOM when uploading large files with S3TransferManager and CRT S3AsyncClient. 
JDK: 17.0.7
OS: CentOS 7
CRT: 0.22.2
AWS SDK Java: 2.20.92
targetThroughputInGbps: default (10Gbps)

The JVM heap dump shows that it stores too many ByteBuffer in ByteBufferStoringSubscriber. 

Attach the heap dump analysis file, hope it helps.
[dump.pdf](https://github.com/aws/aws-sdk-java-v2/files/12208598/dump.pdf)

**Comment by debora-ito** [comment]
Just a quick update: we are actively working on this, still trying to root cause.

**Comment by rtjain21** [comment]
@debora-ito any update on this issue

**Comment by linzhu-isoftstone** [comment]
@debora-ito Any updates on the fix?

**Comment by sigit2** [comment]
Causing our Java app to exhaust its memory, with 1000s of small file transfers.

**Comment by debora-ito** [comment]
We apologize for the long silence, we have some updates to share.

The CRT team made some changes in the CRT client core, released in `aws-crt` 0.29.7, and we observed improvements in the memory usage after running benchmarks using this new version.

We are also exposing a new attribute in the S3CrtAsyncClient that will provide more control over the utilized memory at the client level: `maxNativeMemoryLimitInBytes`. If a value is not provided, the CRT will attempt to limit native memory usage in an optimal way, based on parameters like target throughput. The new attribute was added via https://github.com/aws/aws-sdk-java-v2/pull/4885 and will be available in today's release. 

In summary:
- Please upgrade to SDK version 2.23.5 (which includes `aws-crt` 0.29.7) or greater, and let us know if you see improvements regarding memory usage
- After https://github.com/aws/aws-sdk-java-v2/pull/4885 gets released, please upgrade to the latest version if you want to provide a custom maxNativeMemoryLimitInBytes

**Comment by tosehee** [comment]
I upgraded to 2.25.3 for s3 and 0.29.11 for crt, and the issue still persists.

**Comment by thai-op** [comment]
I'm on 2.23.8 and I still see the issue.

**Comment by vladimirfx** [comment]
Downloading one 3Gb file with concurrency=1, maxNativeMemoryLimitInBytes=2097152 leads to OOMError with native memory usage over 200Mb.

Unfortunately, I can't switch S3 client type because only CRT supports download by parts which is required for downloading big (>1Gb) files.

**Comment by jompox** [comment]
We're also still seeing this issue with the new versions. Could you please re-open this issue?

**Comment by joviegas** [comment]
@jompox  Can you please open a new issue and also mention steps to recreate it
